docs/linking_events.md

```diff
@@ -0,0 +1,97 @@
+
+## Linking multiple events
+
+You can provide an array of events to link to the target stream:
+
+```ruby
+class SomethingHappened < PgEventstore::Event
+end
+
+events = 3.times.map { |i| SomethingHappened.new(type: 'some-event', data: { title: "Something happened-#{i}" }) }
+events_stream = PgEventstore::Stream.new(context: 'FooCtx', stream_name: 'MyAwesomeStream', stream_id: '1')
+projection_stream = PgEventstore::Stream.new(context: 'FooCtx', stream_name: 'MyAwesomeProjection', stream_id: '1')
+events = PgEventstore.client.append_to_stream(events_stream, events)
+# Link events
+PgEventstore.client.link_to(projection_stream, events)
+```
+
+## Handling concurrency
+
+Linking concurrency is implemented the same way as appending concurrency. You can check [**Handling concurrency**](appending_events.md#handling-concurrency) chapter of **Appending Events** section.
+
+Example:
+
+```ruby
+require 'securerandom'
+class SomethingHappened < PgEventstore::Event
+end
+
+event1 = SomethingHappened.new
+event2 = SomethingHappened.new
+
+events_stream = PgEventstore::Stream.new(context: 'FooCtx', stream_name: 'MyAwesomeStream', stream_id: '1')
+projection_stream = PgEventstore::Stream.new(context: 'FooCtx', stream_name: 'MyAwesomeProjection', stream_id: SecureRandom.uuid)
+
+event1, event2 = PgEventstore.client.append_to_stream(events_stream, [event1, event2])
+
+# Links first event
+PgEventstore.client.link_to(projection_stream, event1, options: { expected_revision: :no_stream })
+# Raises PgEventstore::WrongExpectedVersionError error because stream already exists
+PgEventstore.client.link_to(projection_stream, event2, options: { expected_revision: :no_stream })
+```
+
+## Middlewares
+
+If you would like to modify your link events before they get persisted - you should use the `:middlewares` argument which allows you to pass the list of middlewares you would like to use. **By default no middlewares will be applied to the link event despite on `config.middlewares` option**.
+
+**Warning! It is recommended your middlewares do not change `Event#type` and `Event#link_id`.** Otherwise linking feature will simply not work.
+
+Let's say you have these registered middlewares:
+
+```ruby
+PgEventstore.configure do |config|
+  config.middlewares = { foo: FooMiddleware.new, bar: BarMiddleware.new, baz: BazMiddleware.new }
+end
+```
+
+And you want to use `FooMiddleware` and `BazMiddleware`. You simply have to provide an array of corresponding middleware keys you would like to use:
+
+```ruby
+events_stream = PgEventstore::Stream.new(context: 'FooCtx', stream_name: 'MyAwesomeStream', stream_id: '1')
+projection_stream = PgEventstore::Stream.new(context: 'FooCtx', stream_name: 'MyAwesomeProjection', stream_id: '1')
+
+event = PgEventstore.client.append_to_stream(events_stream, PgEventstore::Event.new)
+PgEventstore.client.link_to(projection_stream, event, middlewares: %i[foo baz])
```

Comment:
Good, we can use middleware during the linking process, To be honest, I don't see the use case to modify somehow links. To me make sense of course use middleware on the original link but can't imagine why on the link. 
Ahh.... maybe when we have a link to the event with payload store value and you want to resolve link then you still need middleware to extract value for the resolved link event? Is it true?

Reply:
You are probably mixing linking events and reading events. When linking an event, your middleware will receive next event:

```ruby
Event.new(link_id: original_event.id, type: Event::LINK_TYPE)
```

Currently we don't have any use case where we would modify link event before it gets persisted.